src/Spectre.libGaussianMixtureModelling/MaximizationRunnerRef.h

```diff
@@ -48,81 +48,114 @@ limitations under the License.
+/// algorithm. Serves as a reference to learn from and is purpously not
+/// optimized to avoid obfuscation of the code.
 /// </summary>
 class MaximizationRunnerRef
 {
 public:
     /// <summary>
-    /// Constructor initializing the class with data required during maximization step.
+    /// Constructor initializing the class with data required during
+    /// maximization step.
     /// </summary>
     /// <param name="spectrum">Spectrum to base calculations on.</param>
-    /// <param name="affilationMatrix">Matrix symbolising the probability of affilation
-    /// of each sample to a certain gaussian component.</param>
+    MaximizationRunnerRef(SpectrumView spectrum)
+    {
+        spectrum.mzs.size(); // Needed to avoid warning of unused param.
+    }
+
+    /// <summary>
+    /// Updates weights, means and tandard deviations in a single run.
+    /// </summary>
+    /// <param name="affilationMatrix">Matrix symbolising the probability
+    /// of affilation of each sample to a certain gaussian component.
+    /// </param>
+    /// <param name="spectrum">Spectrum to base calculations on.</param>
     /// <param name="components">Gaussian components to be updated</param>
-    /// <exception cref="ArgumentNullException">Thrown when either of mzArray or intensities pointers are null</exception>
-    MaximizationRunnerRef(SpectrumView spectrum, Matrix<DataType> &affilationMatrix, std::vector<GaussianComponent> &components)
-        : m_Spectrum(spectrum), m_AffilationMatrix(affilationMatrix), m_Components(components)
+    void Maximization(Matrix<DataType> &affilationMatrix,
+        SpectrumView spectrum, GaussianMixtureModel& components)
     {
+        UpdateWeights(affilationMatrix, spectrum, components);
+        UpdateMeans(affilationMatrix, spectrum, components);
+        UpdateStdDeviations(affilationMatrix, spectrum, components);
     }
 
+private:
     /// <summary>
-    /// Updates weights in gaussian components, based on affilation (gamma) matrix.
+    /// Updates weights in gaussian components, based on affilation matrix.
     /// </summary>
-    void UpdateWeights()
+    /// <param name="affilationMatrix">Matrix symbolising the probability
+    /// of affilation of each sample to a certain gaussian component.
+    /// </param>
+    /// <param name="spectrum">Spectrum to base calculations on.</param>
+    /// <param name="components">Gaussian components to be updated</param>
+    void UpdateWeights(Matrix<DataType> &affilationMatrix,
+        SpectrumView spectrum, GaussianMixtureModel& components)
     {
         // This part conducts the instruction:
         // "Using the gamma calculated in the Expectation step, 
         // calculate in the following order, for every k: (phi part)"
         // from the document
         DataType totalDataSize = 0.0;
-        for (unsigned i = 0; i < (unsigned)m_Spectrum.mzs.size(); i++)
+        for (unsigned i = 0; i < (unsigned)spectrum.mzs.size(); i++)
```

Comment:
Wait... Are you computing this each time you do _Maximization_ step? This should never change. What about precomputing this once and not repeating this operation?

Reply:
It is precomputed in the optimized version MaximizationRunnerOpt.h